### PR TITLE
graph: use truncindex consistently to iterate graph entries

### LINKED ISSUE #13992: Truncindex must be able iterate over the existing items and the ability must be used in graph.walkAll()
Now graph.walkAll() does ReadDir() and next trying to do graph.Get() for the each item in the directory. But graph.Get() does graph.idIndex.Get() and if passed ID does not exist in index then graph.Get() returns an error.

So why we need to ReadDir() if we can iterate over graph.idIndex.

### PR DESCRIPTION
**Pull request by rstrlcpy**
Closes #13992
added capability to iterate over all indexes; use the added capability instead of ReadDir() during walking over graph

Signed-off-by: Roman Strashkin roman.strashkin@gmail.com

### CODE REVIEW

**Comment by duglin** [comment]
Please add `Closes #13992` to your first comment so git will close it when this PR is closed.

**Comment by calavera** [comment]
Do you mind to set a more descriptive title? even if it's to close an issue, it helps people to know what's about without having to jump to another page.

**Comment by stevvooe** [comment]
What problem is this actually fixing? This assumes that truncindex is the canonical source of image identifiers.

**Comment by Ramzec** [comment]
Graph.walkAll() calls ReadDir(Graph.root) to retrieve list of IDs.
next Graph.walkAll() calls Graph.Get(id) for each id in ids, that calls Graph.idIndex.Get(id).

So "Yes" trunkindex is the canonical source of image identifiers.

File: pkg/truncindex/truncindex.go [hunk 1]
```diff
@@ -108,3 +108,12 @@ func (idx *TruncIndex) Get(s string) (string, error) {
 	}
 	return "", fmt.Errorf("no such id: %s", s)
 }
+
+func (idx *TruncIndex) Iterate(handler func(id string)) {
```
Comment (stevvooe) [inline pkg/truncindex/truncindex.go]: Please document the method.

**Comment by stevvooe** [comment]
@Ramzec The image IDs are actually loaded from the filesystem, so the canonical source, or the system of record is that graph root. This is kept in sync with the truncindex, which is kept in memory. This happens in `Graph.restore`.

This change is probably okay since we remove the repeated walk code.

LGTM

File: pkg/truncindex/truncindex_test.go [hunk 2]
```diff
@@ -96,6 +96,29 @@ func TestTruncIndex(t *testing.T) {
...
+	index := NewTruncIndex(ids)
+
+	index.Iterate(func(targetId string) {
+		for _, id := range ids {
+			if targetId == id {
+				return
+			}
+		}
+
+		t.Fatal("An unknown ID '%s'", targetId)
```
Comment (duglin) [inline pkg/truncindex/truncindex_test.go]: this needs to be `t.Fatalf`

**Comment by duglin** [comment]
please squash your commits down to one

**Comment by stevvooe** [comment]
@duglin My LGTM is for design review. Need another to go into code review.

**Comment by duglin** [comment]
@stevvooe ok - I edited your LGMT to make it clear its for the design - in case someone misses your clarification.

File: graph/graph.go [hunk 3]
```diff
@@ -341,7 +332,7 @@ func (graph *Graph) ByParent() (map[string][]*image.Image, error) {
 			byParent[parent.ID] = []*image.Image{img}
 		}
 	})
-	return byParent, err
+	return byParent, nil
```
Comment (icecrime) [inline graph/graph.go]: Can we remove the `error` return value if it's always `nil` anyway?

Comment (Ramzec) [inline graph/graph.go]: Make sense. The error can be only in case of problems with reading graph directory.

**Comment by Ramzec** [comment]
rebased with the latest master
removed the returning of error from ByParent

**Comment by calavera** [comment]
code LGTM, but can we remove all the returning errors when there is nothing to return? it looks like `Map` and `Heads` always return nil errors too.

**Comment by Ramzec** [comment]
Cannot understand why the tests failed.

**Comment by stevvooe** [comment]
LGTM

**Pull request merged**